core/chains/solana/orm.go

```diff
@@ -0,0 +1,158 @@
+// TODO: Improve code reuse (mostly c/p of core/chains/terra/orm.go)
+package solana
+
+import (
+	"database/sql"
+
+	"github.com/smartcontractkit/sqlx"
+
+	"github.com/smartcontractkit/chainlink-solana/pkg/solana/db"
+
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services/pg"
+)
+
+type orm struct {
+	q pg.Q
+}
+
+//go:generate mockery --name ORM --srcpkg github.com/smartcontractkit/chainlink-solana/pkg/solana/db --output ./mocks/ --case=underscore
+var _ db.ORM = (*orm)(nil)
+
+// NewORM returns an ORM backed by db.
+func NewORM(db *sqlx.DB, lggr logger.Logger, cfg pg.LogConfig) db.ORM {
+	return &orm{q: pg.NewQ(db, lggr.Named("ORM"), cfg)}
+}
+
+func (o *orm) Chain(id string, qopts ...pg.QOpt) (dbchain db.Chain, err error) {
+	q := o.q.WithOpts(qopts...)
+	chainSQL := `SELECT * FROM solana_chains WHERE id = $1;`
+	err = q.Get(&dbchain, chainSQL, id)
+	return
+}
+
+func (o *orm) CreateChain(id string, config db.ChainCfg, qopts ...pg.QOpt) (chain db.Chain, err error) {
+	q := o.q.WithOpts(qopts...)
+	sql := `INSERT INTO solana_chains (id, cfg, created_at, updated_at) VALUES ($1, $2, now(), now()) RETURNING *`
+	err = q.Get(&chain, sql, id, config)
+	return
+}
+
+func (o *orm) UpdateChain(id string, enabled bool, config db.ChainCfg, qopts ...pg.QOpt) (chain db.Chain, err error) {
+	q := o.q.WithOpts(qopts...)
+	sql := `UPDATE solana_chains SET enabled = $1, cfg = $2, updated_at = now() WHERE id = $3 RETURNING *`
+	err = q.Get(&chain, sql, enabled, config, id)
+	return
+}
+
+func (o *orm) DeleteChain(id string, qopts ...pg.QOpt) error {
+	q := o.q.WithOpts(qopts...)
+	query := `DELETE FROM solana_chains WHERE id = $1`
+	result, err := q.Exec(query, id)
+	if err != nil {
+		return err
+	}
+	rowsAffected, err := result.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if rowsAffected == 0 {
+		return sql.ErrNoRows
+	}
+	return nil
+}
+
+func (o *orm) Chains(offset, limit int, qopts ...pg.QOpt) (chains []db.Chain, count int, err error) {
+	q := o.q.WithOpts(qopts...)
+	if err = q.Get(&count, "SELECT COUNT(*) FROM solana_chains"); err != nil {
+		return
+	}
+
+	sql := `SELECT * FROM solana_chains ORDER BY created_at, id LIMIT $1 OFFSET $2;`
+	if err = q.Select(&chains, sql, limit, offset); err != nil {
```

Comment:
Use a readonly transaction here

Reply:
This is a common pattern throughout the whole repo. Story to follow up here: https://app.shortcut.com/chainlinklabs/story/34583/db-use-txs-for-atomic-paged-queries